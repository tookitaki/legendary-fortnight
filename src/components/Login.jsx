import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Button, Card, Form, Input, Typography } from 'antd';
import { connect } from 'react-redux';
import { loginUser } from '../actions/login';
import { getLoginDetails } from '../selectors/login';
import Logout from './Logout';
import paths from '../constants/path';

const { Text } = Typography;
export const Login = ({ login, dispatchLoginUser, history }) => {
  const { loading, isLoggedIn, error } = login;
  const [state, setState] = useState({
    username: '',
    password: ''
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    const str = `${state.username}:${state.password}`;
    localStorage.setItem('user', state.username);
    dispatchLoginUser(btoa(str));
  };

  useEffect(() => {
    if (isLoggedIn) {
      history.push(paths.dashboard);
    }
  }, [isLoggedIn, history]);

  const onChange = (key, e) => {
    state[key] = e.target.value;
    setState(state);
  };

  return (
    <div>
      <div style={{ width: '100%' }}>
        <p style={HeaderStyle}>Login</p>
      </div>
      <Card style={LoginCardStyle}>
        <Form layout="vertical" onSubmit={handleSubmit}>
          <Form.Item label="Email">
            <Input
              id="username"
              data-testid="username"
              type="text"
              placeholder="e.g. example@tookitaki.com"
              onChange={(e) => onChange('username', e)}
            />
          </Form.Item>
          <Form.Item label="Password">
            <Input
              id="password"
              data-testid="password"
              type="password"
              placeholder="Must contain 8 characters"
              onChange={(e) => onChange('password', e)}
            />
          </Form.Item>
          <Form.Item style={{ textAlign: 'center' }}>
            <Button
              id="login"
              data-testid="login"
              disabled={loading}
              type="primary"
              htmlType="submit"
              loading={loading}
              block
              onClick={handleSubmit}>
              Login
            </Button>
            <Logout loading={loading} />
          </Form.Item>
        </Form>
        {error && (
          <Text type="danger">
            {error.data || error.message || String(error)}
          </Text>
        )}
      </Card>
    </div>
  );
};

const LoginCardStyle = {
  width: '500px',
  marginLeft: 'auto',
  marginRight: 'auto',
  marginTop: '35px',
  border: '1px solid #D0D7E3',
  borderRadius: '4px',
  padding: '10px'
};

const HeaderStyle = {
  marginLeft: 'auto',
  marginRight: 'auto',
  marginTop: '15px',
  display: 'block',
  fontStyle: 'normal',
  fontWeight: '600',
  fontSize: '16px',
  color: '#36465E',
  width: '100%',
  textAlign: 'center'
};

Login.propTypes = {
  dispatchLoginUser: PropTypes.func,
  login: PropTypes.shape({
    loading: PropTypes.bool,
    isLoggedIn: PropTypes.bool,
    error: PropTypes.string
  }),
  history: PropTypes.func
};

const mapStateToProps = (state) => {
  return {
    login: getLoginDetails(state)
  };
};

const mapDispatchToProps = {
  dispatchLoginUser: loginUser
};
export default connect(mapStateToProps, mapDispatchToProps)(Login);
