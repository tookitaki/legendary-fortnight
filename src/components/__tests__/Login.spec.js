import Login from '../Login';
import history from '../../utils/history';
import { fireEvent, render } from '@testing-library/react';

//mocking other components
jest.mock('../Logout', () => () => <div>Logout</div>);

describe('<Login />', () => {
  let props, component;
  beforeEach(() => {
    //Manual Mock 'window.matchMedia'
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn()
      }))
    });
    history.push = jest.fn();
    props = {
      login: {
        loading: false,
        isLoggedIn: false,
        error: null
      },
      dispatchLoginUser: jest.fn(),
      dispatchLogoutUser: jest.fn()
    };
    component = render(<Login {...props} />);
  });

  afterEach(() => {
    history.push.mockClear();
  });

  it('should render login snapshot', () => {
    expect(component.container).toMatchSnapshot();
  });

  it('should call login on clieck of LOGIN button', () => {
    const { getByTestId } = component;

    const username = getByTestId('username');
    fireEvent.change(username, { target: { value: 'test-username' } });

    const password = getByTestId('password');
    fireEvent.change(password, { target: { value: 'test-password' } });

    const login = getByTestId('login');
    fireEvent.click(login, { preventDefault: jest.fn() });

    const str = `test-username:test-password`;
    expect(props.dispatchLoginUser).toHaveBeenCalledWith(btoa(str));
  });
});
