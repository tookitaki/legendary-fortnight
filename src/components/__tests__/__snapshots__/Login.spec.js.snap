// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render login snapshot 1`] = `
<div>
  <div>
    <div
      style="width: 100%;"
    >
      <p
        style="margin-left: auto; margin-right: auto; margin-top: 15px; display: block; font-style: normal; font-weight: 600; font-size: 16px; color: rgb(54, 70, 94); width: 100%; text-align: center;"
      >
        Login
      </p>
    </div>
    <div
      class="ant-card ant-card-bordered"
      style="width: 500px; margin-left: auto; margin-right: auto; margin-top: 35px; border: 1px solid #d0d7e3; border-radius: 4px; padding: 10px;"
    >
      <div
        class="ant-card-body"
      >
        <form
          class="ant-form ant-form-vertical"
        >
          <div
            class="ant-row ant-form-item"
          >
            <div
              class="ant-col ant-form-item-label"
            >
              <label
                class=""
                title="Email"
              >
                Email
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input"
                    data-testid="username"
                    id="username"
                    placeholder="e.g. example@tookitaki.com"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-row ant-form-item"
          >
            <div
              class="ant-col ant-form-item-label"
            >
              <label
                class=""
                title="Password"
              >
                Password
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input"
                    data-testid="password"
                    id="password"
                    placeholder="Must contain 8 characters"
                    type="password"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-row ant-form-item"
            style="text-align: center;"
          >
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <button
                    class="ant-btn ant-btn-primary ant-btn-block"
                    data-testid="login"
                    id="login"
                    type="submit"
                  >
                    <span>
                      Login
                    </span>
                  </button>
                  <div>
                    Logout
                  </div>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
